### PR TITLE
CA-80, CA-82, CA-84

### PR DESCRIPTION
**Pull request by marcbeermann**
CA-80 StartProjectViewModel, CA-82 DisplayProjectViewModel, CA-84 BrowseProjectsViewModel

- The Create Project Action actually uses the new EditProjectViewModel (as does the Edit Project Action).
- Added validation when user attempts to create a project of the same name.
- Made changes to ProjectsController to calculate the participation counts (needed for 'Progress' field)
- Added 'RemainingDays' and 'StatusDescription' properties to the Project model.

### CODE REVIEW

File: src/CollAction/Controllers/ProjectsController.cs [hunk 1]
```diff
@@ -59,13 +61,30 @@ public ViewResult StartInfo()
...
+            if (items.Count == 0)
             {
                 return NotFound();
             }
 
-            return View(project);
+            return View(items.First());
+        }
+
+        public async Task<List<DisplayProjectViewModel>> GetProjectDisplayItemsWhere(Expression<Func<Project, bool>> WhereExpression)
```
Comment (stewartmatheson) [inline src/CollAction/Controllers/ProjectsController.cs]: Does this method belong in the controller? Coming from a rails background I would look to express this somewhere else but if this is an ASP.NET thing to do then that's fine. Don't let it block the pull request.

**Comment by stewartmatheson** [comment]
Oh also I would have one pull request per ticket as well as one ticket per commit. Makes history easier to understand down the track. Again should not block anything just something worth noting.

File: src/CollAction/Models/Project.cs [hunk 2]
```diff
@@ -67,11 +67,32 @@ public bool IsClosed
...
         public string HashTags
             => string.Join(";", Tags?.Select(tag => tag.Tag.Name) ?? Enumerable.Empty<string>());
 
+        [NotMapped]
+        public int RemainingDays
+            => Convert.ToInt32(Math.Round((End - Start).TotalDays));
+
+        [NotMapped]
+        [Display(Name = "Status")]
+        public string StatusDescription
```
Comment (logophobia) [inline src/CollAction/Models/Project.cs]: A more appropiate place would be to put it in the viewmodel. It's purely related to how to display a project, not the project data itself. It's also not localized. In fact, if you put this property here, it would also be impossible to localize in the future.

File: src/CollAction/Models/ProjectViewModels/DisplayProjectViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations;
+
+namespace CollAction.Models
+{
+    public class DisplayProjectViewModel
+    {
+        public Project Project;
```
Comment (logophobia) [inline src/CollAction/Models/ProjectViewModels/DisplayProjectViewModel.cs]: A small style issue, use:
public Project Project { get; set; }
for this, a shorthand for a getter. If you make it a property, then it's possible to change the implementation and storage of the property without changing the interface.

It doesn't matter terribly much for viewmodel classes, still, good habit to get into. https://msdn.microsoft.com/en-us/library/x9fsa0sw.aspx

File: src/CollAction/Models/ProjectViewModels/DisplayProjectViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+        public int Participants;
```
Comment (logophobia) [inline src/CollAction/Models/ProjectViewModels/DisplayProjectViewModel.cs]: Same as above.

File: src/CollAction/Controllers/ProjectsController.cs [hunk 5]
```diff
@@ -34,21 +35,22 @@ public ViewResult StartInfo()
         }
 
         // GET: Project/Find
-        public async Task<IActionResult> Find(FindProjectViewModel model)
+        public async Task<IActionResult> Find(FindProjectsViewModel model)
         {
             if (model.SearchText == null) {
-                return View(new FindProjectViewModel { Projects = await _context.Projects.ToListAsync() });
+                return View(new FindProjectsViewModel {
+                    Projects = await GetProjectDisplayItemsWhere(p => p.Status != ProjectStatus.Hidden)
```
Comment (logophobia) [inline src/CollAction/Controllers/ProjectsController.cs]: Status deleted also has to be filtered

File: src/CollAction/Controllers/ProjectsController.cs [hunk 6]
```diff
@@ -34,21 +35,22 @@ public ViewResult StartInfo()
...
+                });
             }
 
-            model.Projects = await _context.Projects.Where(p => p.Name.Contains(model.SearchText) || p.Description.Contains(model.SearchText) || p.Goal.Contains(model.SearchText)).ToListAsync();
-            
+            model.Projects = await GetProjectDisplayItemsWhere(
+                p => p.Status != ProjectStatus.Hidden && (p.Name.Contains(model.SearchText) || p.Description.Contains(model.SearchText) || p.Goal.Contains(model.SearchText)));
```
Comment (logophobia) [inline src/CollAction/Controllers/ProjectsController.cs]: Filter status deleted

File: src/CollAction/Models/Project.cs (hunk 2 above)
Comment (marcbeermann) [inline src/CollAction/Models/Project.cs]: That makes a lot of sense, I've done another commit with this moved to the DisplayProjectViewModel.

File: src/CollAction/Views/Projects/Create.cshtml [hunk 7]
```diff
@@ -2,17 +2,17 @@
 
 @inject IViewLocalizer Localizer
 
-@model CollAction.Models.CreateProjectViewModel
```
Comment (logophobia) [inline src/CollAction/Views/Projects/Create.cshtml]: Right now CreateProject and EditProject use the same data. I can imagine it'd be temping to unify the viewmodels, but this feels rather premature.

I'm almost certain that in the future we'll do things slightly different. A project that's being editted for example:
* Might not be allowed to change the start date retroactively
* Might be able to administrate comments, block unwanted users (just fictional examples). Stuff that's not relevant to CreateProject

Tying them together means changing something in the edit workflow, might break create, and retrospectively.

Trying to unify stuff like this should be done at the end of the project, when we know more about how the final project looks.

File: src/CollAction/Views/Projects/Index.cshtml [hunk 8]
```diff
@@ -47,33 +53,39 @@
...
             </td>
             <td>
-                @Html.DisplayFor(content => item.End)
+                @Html.DisplayFor(content => item.Progress)%
             </td>
             <td>
-                <a asp-action="Edit" asp-route-id="@item.Id">Edit</a> |
-                <a asp-action="Details" asp-route-id="@item.Id">Details</a> |
-                <a asp-action="Delete" asp-route-id="@item.Id">Delete</a>
+                <a asp-action="Edit" asp-route-id="@item.Project.Id">@Localizer["Edit"]</a> |
```
Comment (logophobia) [inline src/CollAction/Views/Projects/Index.cshtml]: This is still stuff left over from the auto-generated project views that asp.net can generate for you. I don't think it makes a lot of sense to show edit and delete buttons to everyone in all the project lists.

File: src/CollAction/Controllers/ProjectsController.cs [hunk 9]
```diff
@@ -84,21 +103,21 @@ public ViewResult StartInfo()
         [HttpPost]
         [ValidateAntiForgeryToken]
         [Authorize]
-        public async Task<IActionResult> Create([Bind("Name,Description,Goal,CategoryId,LocationId,Target,End")] CreateProjectViewModel createProjectVM)
+        public async Task<IActionResult> Create([Bind("Name,Description,Goal,CategoryId,LocationId,Target,End")] EditProjectViewModel createProjectViewModel)
```
Comment (logophobia) [inline src/CollAction/Controllers/ProjectsController.cs]: EditProjectViewModel createProjectViewModel

Another issue with conflating the create and edit workflows

File: src/CollAction/Controllers/ProjectsController.cs [hunk 10]
```diff
@@ -160,9 +184,15 @@ public ViewResult StartInfo()
         [HttpPost]
         [ValidateAntiForgeryToken]
         [Authorize]
-        public async Task<IActionResult> Edit(int id, [Bind("Id,Name,Description,Goal,CategoryId,LocationId,Target,End")] Project project)
+        public async Task<IActionResult> Edit(int id, [Bind("Id,OwnerId,Name,Description,Goal,CategoryId,LocationId,Target,End")] EditProjectViewModel editProjectViewModel)
```
Comment (logophobia) [inline src/CollAction/Controllers/ProjectsController.cs]: Here you're taking the owner id as parameter. This means the user could change the owner-id from the form, and circumvent the owner validation! Security vulnerability.

File: src/CollAction/Controllers/ProjectsController.cs [hunk 11]
```diff
@@ -110,23 +129,16 @@ public ViewResult StartInfo()
                 }
                 catch (DbUpdateException ex)
                 {
-                    if (ex.InnerException is PostgresException)
```
Comment (logophobia) [inline src/CollAction/Controllers/ProjectsController.cs]: * Validation this way is probably most efficient, but I really would rather put validation like this either in the viewmodel or just as a separate query at the start (AnyAsync(project => project.Name.Equals..)). I realize that might lead to rare race conditions (two people submitting projects with the same same at the exact same time), but it's more important that validation (exception code checking isn't really self-explanatory) is clear rather than a small gain in efficiency in a workflow that'd only be used when creating projects. Worst case scenario, one in a million people get a generic, error creating project, and retry. But our validation is a lot more clear.
* In the future you could use exception filters for stuff where the catch(ExceptionType.. isn't sufficient

**Comment by logophobia** [comment]
Overall, good stuff. Does need some tweaks.
